src/config.rs

```diff
@@ -0,0 +1,452 @@
+
+use std::fs::File;
+use std::io::{Read, Error};
+use std::path::{PathBuf, Path};
+use std::default::Default;
+use std::fmt;
+use std::collections::{HashMap, HashSet, BTreeMap, BTreeSet};
+use std::hash::Hash;
+use std::cmp::Eq;
+
+use yaml_rust::{Yaml, YamlLoader, ScanError};
+
+pub enum ConfigError {
+    YamlScan(ScanError),
+    YamlParse(ConfigMeta),
+    FileError(String, Error),
+    MissingExternalFile(ConfigMeta),
+}
+
+impl ConfigError {
+    pub fn to_string(&self) -> String {
+        match self {
+            &ConfigError::YamlScan(ref e) => format!("Failed to scan YAML: {}", e),
+            &ConfigError::YamlParse(ref meta) => {
+                let mut path = String::new();
+
+                for (index, element) in meta.fields.iter().enumerate() {
+                    if index != 0 {
+                        path = path + "->";
+                    }
+
+                    path = path + element;
+                }
+
+                let message = if meta.bad_value {
+                    "Could not find YAML"
+                } else {
+                    "Failed to parse YAML"
+                };
+
+                let basic = format!("{}: {}: {}: expected {}", meta.path.display(), message, path, meta.ty);
+
+                let options = if meta.options.len() > 0 {
+                    let mut result = "".to_string();
+
+                    for (index, option) in meta.options.iter().enumerate() {
+                        if index != 0 {
+                            result = result + ", ";
+                        }
+
+                        result = result + option;
+                    }
+
+                    format!("\n{}:\t {} {{ {} }}", meta.path.display(), meta.ty, result)
+                } else {
+                    "".to_string()
+                };
+
+                format!("{}{}", basic, options)
+            },
+            &ConfigError::FileError(ref disp, ref e) => format!("Config File Error: \"{}\", {}", disp, e),
+            &ConfigError::MissingExternalFile(ref meta) => format!("{}: External YAML file is missing", meta.path.display()),
+        }
+    }
+}
+
+impl fmt::Debug for ConfigError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.to_string())
+    }
+}
+
+impl fmt::Display for ConfigError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.to_string())
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct ConfigMeta {
+    path: PathBuf, // Where the file is located, "" if not from a file.

```

Comment:
Decided to make it default to "config\\config.yml" instead as a default directory for amethyst.